### PR TITLE
[Fluent][Redy for review] Graph.RBAC (basic) and Key Vault

### PR DESCRIPTION
**Pull request by jianghaolu**

### CODE REVIEW

**Comment by hovsepm** [comment]
> using System;

Please add license header

---
Refers to: src/ResourceManagement/Graph.RBAC/Microsoft.Azure.Management.Fluent.Graph.RBAC/GraphRBACManager.cs:1 in 1690c7b. [](commit_id = 1690c7b81eb5f47d51b6d4fb842d340263faf263, deletion_comment = False)

File: src/ResourceManagement/Graph.RBAC/Microsoft.Azure.Management.Fluent.Graph.RBAC/ServicePrincipalImpl.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
+/**
+* Copyright (c) Microsoft Corporation. All rights reserved.
+* Licensed under the MIT License. See License.txt in the project root for
+* license information.
+*/
```
Comment (hovsepm) [inline src/ResourceManagement/Graph.RBAC/Microsoft.Azure.Management.Fluent.Graph.RBAC/ServicePrincipalImpl.cs]: please use C# style license header

File: src/ResourceManagement/Graph.RBAC/Microsoft.Azure.Management.Fluent.Graph.RBAC/ServicePrincipalsImpl.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
+/**
+* Copyright (c) Microsoft Corporation. All rights reserved.
+* Licensed under the MIT License. See License.txt in the project root for
+* license information.
+*/ 
+
```
Comment (hovsepm) [inline src/ResourceManagement/Graph.RBAC/Microsoft.Azure.Management.Fluent.Graph.RBAC/ServicePrincipalsImpl.cs]: C# style

File: src/ResourceManagement/Graph.RBAC/Microsoft.Azure.Management.Fluent.Graph.RBAC/UserImpl.cs [hunk 3]
```diff
@@ -0,0 +1,145 @@
+/**
+* Copyright (c) Microsoft Corporation. All rights reserved.
+* Licensed under the MIT License. See License.txt in the project root for
+* license information.
+*/
```
Comment (hovsepm) [inline src/ResourceManagement/Graph.RBAC/Microsoft.Azure.Management.Fluent.Graph.RBAC/UserImpl.cs]: C# Style

File: src/ResourceManagement/Graph.RBAC/Microsoft.Azure.Management.Fluent.Graph.RBAC/UsersImpl.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
+/**
+* Copyright (c) Microsoft Corporation. All rights reserved.
+* Licensed under the MIT License. See License.txt in the project root for
+* license information.
+*/ 
+
```
Comment (hovsepm) [inline src/ResourceManagement/Graph.RBAC/Microsoft.Azure.Management.Fluent.Graph.RBAC/UsersImpl.cs]: C# Style

File: src/ResourceManagement/Graph.RBAC/Microsoft.Azure.Management.Fluent.Graph.RBAC/Domain/IActiveDirectoryGroup.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
+/**
+* Copyright (c) Microsoft Corporation. All rights reserved.
+* Licensed under the MIT License. See License.txt in the project root for
+* license information.
+*/ 
+
```
Comment (hovsepm) [inline src/ResourceManagement/Graph.RBAC/Microsoft.Azure.Management.Fluent.Graph.RBAC/Domain/IActiveDirectoryGroup.cs]: C# Style

**Comment by hovsepm** [comment]
> using System;

Please add license header

---
Refers to: src/ResourceManagement/KeyVault/Microsoft.Azure.Management.Fluent.KeyVault/KeyVaultManager.cs:1 in 1690c7b. [](commit_id = 1690c7b81eb5f47d51b6d4fb842d340263faf263, deletion_comment = False)

**Comment by hovsepm** [comment]
> using Microsoft.IdentityModel.Clients.ActiveDirectory;

Please add license header

---
Refers to: src/ResourceManagement/Resource/Microsoft.Azure.Management.V2.Resource/Authentication/AzureCredentials.cs:1 in 1690c7b. [](commit_id = 1690c7b81eb5f47d51b6d4fb842d340263faf263, deletion_comment = False)

**Comment by hovsepm** [comment]
you will need to pull from upstream and resolve conflicts

File: src/ResourceManagement/Graph.RBAC/Microsoft.Azure.Management.Fluent.Graph.RBAC/ServicePrincipalsImpl.cs [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    using Management.Graph.RBAC;
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using Rest.Azure;
+
+    /// <summary>
+    /// The implementation of ServicePrincipals and its parent interfaces.
+    /// </summary>
+    public partial class ServicePrincipalsImpl  :
```
Comment (hovsepm) [inline src/ResourceManagement/Graph.RBAC/Microsoft.Azure.Management.Fluent.Graph.RBAC/ServicePrincipalsImpl.cs]: >public [](start = 4, length = 6)

this one as well - internal

File: src/ResourceManagement/Graph.RBAC/Microsoft.Azure.Management.Fluent.Graph.RBAC/UserImpl.cs [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Management.Graph.RBAC;
+    using V2.Resource.Core.ResourceActions;
+    using System;
+
+    /// <summary>
+    /// Implementation for User and its parent interfaces.
+    /// </summary>
+    public partial class UserImpl :
```
Comment (hovsepm) [inline src/ResourceManagement/Graph.RBAC/Microsoft.Azure.Management.Fluent.Graph.RBAC/UserImpl.cs]: >public [](start = 4, length = 6)

internal

File: src/ResourceManagement/Graph.RBAC/Microsoft.Azure.Management.Fluent.Graph.RBAC/UsersImpl.cs [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    using Microsoft.Azure.Management.V2.Resource.Core.CollectionActions;
+    using Microsoft.Azure.Management.V2.Resource.Core;
+    using System.Threading.Tasks;
+    using Management.Graph.RBAC;
+    using System;
+
+    /// <summary>
+    /// The implementation of Users and its parent interfaces.
+    /// </summary>
+    public partial class UsersImpl :
```
Comment (hovsepm) [inline src/ResourceManagement/Graph.RBAC/Microsoft.Azure.Management.Fluent.Graph.RBAC/UsersImpl.cs]: >public [](start = 4, length = 6)

internal

File: src/ResourceManagement/KeyVault/Microsoft.Azure.Management.Fluent.KeyVault/AccessPolicyImpl.cs [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+    using Microsoft.Azure.Management.V2.Resource.Core.ChildResource.Definition;
+    using Microsoft.Azure.Management.Fluent.KeyVault.AccessPolicy.Update;
+    using System;
+    using Microsoft.Azure.Management.Fluent.KeyVault.Vault.Definition;
+    using V2.Resource.Core.ChildResourceActions;
+
+    /// <summary>
+    /// Implementation for AccessPolicy and its parent interfaces.
+    /// </summary>
+    public partial class AccessPolicyImpl  :
```
Comment (hovsepm) [inline src/ResourceManagement/KeyVault/Microsoft.Azure.Management.Fluent.KeyVault/AccessPolicyImpl.cs]: internal

File: src/ResourceManagement/KeyVault/Microsoft.Azure.Management.Fluent.KeyVault/VaultImpl.cs [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+    using System.Linq;
+    using Microsoft.Azure.Management.V2.Resource;
+    using Management.KeyVault;
+    using Graph.RBAC;
+    using System;
+
+    /// <summary>
+    /// Implementation for Vault and its parent interfaces.
+    /// </summary>
+    public partial class VaultImpl  :
```
Comment (hovsepm) [inline src/ResourceManagement/KeyVault/Microsoft.Azure.Management.Fluent.KeyVault/VaultImpl.cs]: >public [](start = 4, length = 6)

internal

File: src/ResourceManagement/KeyVault/Microsoft.Azure.Management.Fluent.KeyVault/VaultsImpl.cs [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    using Management.KeyVault;
+    using Graph.RBAC;
+    using System;
+    using System.Threading.Tasks;
+    using System.Threading;
+
+    /// <summary>
+    /// The implementation of Vaults and its parent interfaces.
+    /// </summary>
+    public partial class VaultsImpl  :
```
Comment (hovsepm) [inline src/ResourceManagement/KeyVault/Microsoft.Azure.Management.Fluent.KeyVault/VaultsImpl.cs]: >public [](start = 4, length = 6)

internal

**Comment by hovsepm** [comment]
> using Microsoft.IdentityModel.Clients.ActiveDirectory;

Found another one :) license header please! :)

---
Refers to: src/ResourceManagement/Resource/Microsoft.Azure.Management.V2.Resource/Authentication/AzureCredentials.cs:1 in 84bb4a5. [](commit_id = 84bb4a553d9f2f61394d8c58ae83f605245bfc39, deletion_comment = False)

File: src/ResourceManagement/Resource/Microsoft.Azure.Management.V2.Resource/Authentication/AzureCredentials.cs [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+        private string username, password;
+        private string clientId, clientSecret;
+        private string tenantId;
+        private AzureEnvironment environment;
+        private Func<DeviceCodeResult, bool> deviceCodeHandler;
+
+        public TokenCache TokenCache { get; private set; }
+        public string DefaultSubscriptionId { get; private set; }
+
+        private AzureCredentials() {
```
Comment (hovsepm) [inline src/ResourceManagement/Resource/Microsoft.Azure.Management.V2.Resource/Authentication/AzureCredentials.cs]: >{ [](start = 35, length = 1)

this is Java code. { should go to a new line

File: src/ResourceManagement/Resource/Microsoft.Azure.Management.V2.Resource/Authentication/AzureCredentials.cs [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Management.V2.Resource.Authentication
+{
+    public class AzureCredentials : ServiceClientCredentials
```
Comment (hovsepm) [inline src/ResourceManagement/Resource/Microsoft.Azure.Management.V2.Resource/Authentication/AzureCredentials.cs]: please add documentation for the class

File: src/ResourceManagement/Resource/Microsoft.Azure.Management.V2.Resource/Authentication/AzureCredentials.cs [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+            string url = request.RequestUri.ToString();
+            if (url.StartsWith(environment.GraphEndpoint, StringComparison.OrdinalIgnoreCase))
+            {
+                adSettings.TokenAudience = new Uri(environment.GraphEndpoint);
+            }
+            Task<ServiceClientCredentials> credentials;
+            if (username != null && password != null)
+            {
+                credentials = UserTokenProvider.LoginSilentAsync(clientId, tenantId, username, password, adSettings, TokenCache);
+            }
```
Comment (hovsepm) [inline src/ResourceManagement/Resource/Microsoft.Azure.Management.V2.Resource/Authentication/AzureCredentials.cs]: this way you will hit AD Endpoint for each Send and re-Authenticate. Please take a look to TestFramework authentication flow in the sdk-for-net repo.

**Comment by jianghaolu** [comment]
Verified that it's capable of refreshing tokens.

**Comment by hovsepm** [comment]
:shipit:

**Pull request merged**